portfolio/src/main/webapp/script.js

```diff
@@ -29,100 +28,201 @@ function addRandomGreeting() {
+    big = big.toLocaleLowerCase();
 
-//js is pass by value so no need for a copy
-function getClosestTag(index, source){
+    var smallLen = small.length;
+    if (smallLen == 0) {return [];}
 
-    while (source.charAt(index) != '>'){
-        if (index == 0){
-            alert("index hit 0");
-            break;
-        }
-        index--;
+    var startIndex = 0, index, indices = [];
+
+    while ((index = big.indexOf(small, startIndex)) > -1) {
+        indices.push(index);
+        startIndex = index + smallLen;
     }
 
-    var endIndex = index+1; //plus 1 to account for bracket
+    return indices;
+}
 
-    while (source.charAt(index) != '<'){
-        index--;
+/**
+ * Walks the tree and finds matches,
+ * returns index array of matches.
+ * Will optimize to gather straight from tree
+ * rather than push into new arr to save space
+ */
+function treeWalker(node, input){
+    var all = [];
+
+    for (node= node.firstChild; node ; node=node.nextSibling){
+        if (node.nodeType==3){
+            all.push(node);
+        } else {
+            all = all.concat(treeWalker(node, input));
+        }
     }
 
-    var startIndex = index;
+    //find matches
+    var indexArr = [];
+    for (var i = 0; i < all.length; i++){
+        const regex = new RegExp(input, "i");
+        if (regex.test(all[i].textContent) == true){
+            indexArr.push(all[i]);
+        }
+    }
 
-    return source.substring(startIndex, endIndex)
+    return indexArr;
+}
+
+/**
+ * Clears the current highlighted text nodes
+ * Will optimize to actually remove the span rather than the class
+ */
+function clear(){
+
+    var highlightedList = document.querySelectorAll('.highlight');
 
+    highlightedList.forEach(function(el) {
+        el.classList.remove('highlight');
+    });
+    
 }
 
-function retrieveId(string, source){
+/**
+ * Handles clearing old nodes and end of search run,
+ * as well as calls the highlighting and scrolling fns
+ */
+function treeWalkerHelper(matches, input, index){
 
-    var searchForId = string.search("id=\"");
-    if (!searchForId){
+    if (index === matches.length && index != 0){
+        clear();
+        setHidden('nextbutton', true);
+        document.body.scrollIntoView({behavior: "smooth"}); //back to the top
         return;
     }
-    const startIndex = searchForId+4;
-    var retString = string.charAt(startIndex);
 
-    for(var i = startIndex+1; string.charAt(i) != "\""; i++){
-        retString += string.charAt(i);
+    if (index > 0){
+        clear();
     }
 
-    return retString;
+    console.log(matches[index]);
+    highlighter(matches[index], input);
+
+    var els = document.getElementsByClassName('highlight');
+    els[0].scrollIntoView({behavior: "smooth"}); 
 }
 
-function runSearch(input){
+/**
+ * Highlights a given text node
+ */
+function highlighter(node, input){
 
-    var source = retrieveHtml();
+    try {
+        var foundIndices = getIndicesOf(input, node.textContent);
+    } catch(err) {
```

Comment:
That is a good idea. The error comes up every second search, as well as every next call there after. 

Uncaught TypeError: Cannot read property 'textContent' of undefined
    at highlighter (script.js:191)
    at treeWalkerHelper (script.js:174)
    at HTMLInputElement.<anonymous> (script.js:223)

It is happening because the node getting passed into the function is undefined, but the weird thing is the search will still work and highlight the values it needs to regardless. I also think this is contributing to another bug that will miss some words. This is the main thing that I have to figure out.